### DISCUSSION TITLE
Understanding Salt with Hashicorp Vault

### DISCUSSION THREAD

**Discussion by elwood218** [saltstack/salt#67029] [General]
How Vault is integrated into Vault?

According to following documentation I understand this example as all Minions need access to Vault, correct? -> https://salt-extensions.github.io/saltext-vault/topics/basic_configuration.html

But I don't want allow any Minion to access Vault. I want that only the Masters are allowed to access Vault and then I would write pillar files where I access a secret with:
```
secret: '{{ salt.vault.read_secret("secret/data/foo", key="bar") }}'
```

But for this I also have configured it like in the basic configuration on the master at least. But I did not put the right to create vault app-roles. For me it should work like that:
1. I put once a secret-id (most at startup or if the service was longer down and could not renew its token)
2. The secret id will than get a token where it is allowed to read certain secrets and then the masters should renew their token all the time until for any reason (like reboot) it cannot renew in a certain time. Then I would put a new secret-id - but which can be only used once.

BUT obviusly it doesn't work like that. At least for me. Can someone explain how that should be done because the documentation unfortunately does not explain it and the different options.

PS: I also would let master read all secrets and then use that as Pillar directly for the minions if that is possible without let each minion access the vault?! But I wonder then how I could tell that this minion is only allowed to read this certain data. With the way above I can at least tell in top.sls.

Kind regards
Mathias

**Comment by elwood218** [comment]
If it is not working like that I would probably do a workaround and install Vault Agent as proxy and let the masters talk to the local Vault agent with a token which never expires. (at least in theory..)

But maybe it would work because at the moment I still got an error when I try it like menioned above but don't know why. If I use Approle with other Software it works. It tells me following error and I don't know why it would create a token other than I defined in the approle..
```
salt.exceptions.SaltRenderError: Problem running salt function in Jinja template: Failed to read secret! CommandExecutionError: {'error': 'VaultInvocationError: child policies must be subset of parent'}
```

As I am quite new to salt I need to check how I can debug that and see what policies are tried..

**Comment by dmurphy18** [comment]
FYI use the Salt Vault Extension, it is much more up to date, and that is the way Salt is heading with 3008, see https://github.com/salt-extensions/saltext-vault

**Reply by elwood218** [reply]
@dmurphy18 Thank you. That is exactly what I am referring to. The link of the "basic config" is the one on that github page you posted. But unfortunately I don't understand it completely. I don't see why the master must be able to create approles for the minions.. but I guess this would only be needed if the minion also has access to Vault. That is what I still try to figure out.

**Reply by elwood218** [reply]
I have debugged the request and found the problem. So it is reading at least now. The problem was that there are default policies which are set. So I have set my policies which also the secret-id had - so it then was a "subset". This had to be set in master config.
```
vault:
  policies:
    assign:
      - policy1
      - policy2
```
But I still don't know how to use that in the correct way and how is the workflow and if the token is renewed in the background because I don't want to put a secret-id which runs forever.

And @dmurphy18 you probably wanted to tell me that I shall open an discussion in their github repo because it is an extension and not included anymore, right?! :)

**Comment by lkubb** [comment]
> But I don't want allow any Minion to access Vault. I want that only the Masters are allowed to access Vault

The Vault integration works under the following assumptions:

1) The master orchestrates minion authentication/authorization only, it does not have access to any secrets.
2) Minions request Vault server configuration + authentication credentials and access Vault directly. The master assigns [Vault policies](https://developer.hashicorp.com/vault/docs/concepts/policies) (collections of related authorizations you need to define yourself) to these credentials based on the `vault:policies:assign` configuration, which is usually [templated](https://salt-extensions.github.io/saltext-vault/topics/templating.html) because otherwise, all minions would receive the same permissions.
3) A practical exception to (1) is minion pillar compilation, where the master impersonates each minion by issuing minion-specific credentials to itself. This avoids distributing secrets to minions they are not supposed to have access to.

I'm pretty sure omitting the `peer_run` master configuration (which allows minions to request credentials) also breaks (3), so there is currently no way of avoiding minions requesting Vault configuration/credentials. You can still block connections on the network layer.

This fundamental architecture is not going to change. It would be fairly straightforward to disallow non-impersonated authentication requests though, for which you can create a feature request at https://github.com/salt-extensions/saltext-vault/issues.

Note that pillar compilation can be a performance bottleneck for the master and minions only receive access to the secrets they are supposed to be able to read, so in general it's recommended for minions to do the lookups themselves. Be absolutely sure that the described behavior is what you need and be aware of the tradeoffs.

> I would write pillar files where I access a secret

It's not recommended to access the Vault modules during regular `.sls` pillar compilation because of circular reference issues, unless you're not using the pillar to assign policies to your minions (which one generally should do). The [dedicated pillar module](https://salt-extensions.github.io/saltext-vault/ref/pillar/saltext.vault.pillar.vault.html) does not have this limitation.

> But I wonder then how I could tell that this minion is only allowed to read this certain data

A minion is allowed access to any paths its policies allow it to, this question is not specific to the Vault integration. You can show policies a minion receives based on your master configuration via `salt-run vault.show_policies <minion_id>`.

> For me it should work like that [...]

In general, one would add a token/secret ID without a cap on the number of uses, but with one on the validity time. When using Salt 3007.x or the extension, the master automatically renews the current token during requests. While it does not have a dedicated renewal background process, the token is usually kept alive just by the base activity.

Afaik you can use the `env` SDB module to source the `VAULT_TOKEN` from the Vault Agent (https://salt-extensions.github.io/saltext-vault/ref/configuration.html#token) (Edit: This might be broken though - if you decide to test it and it does not work, feel free to submit a bug report).

> Can someone explain how that should be done

The common setups are documented in the guide you linked. I hope this post helps explain the concepts, but feel free to ask any further questions.

**Reply by elwood218** [reply]
@lkubb Really thank you for that fast answer and that detailed explanation and clarification of the concept!
In my view it would be better if only the master would need access to the vault but I am not that deep into the whole Salt so probably as you said the performance could be later a problem and maybe it is also more secure if each minion by itself can access it. I thought at least it would be better to limit that access and I was also not sure if each "minion" also if there can no minion be installed can make use of that if "master" should not be allowed to read secrets.

I didn't want to put secrets on master pillar directly as they are not encrypted and the secrets are in the vault anyway - so there is no need to put and manage secrets in multiple stores. That's why I liked the idea of integrate Vault.

I will test that setup then with access by each minion. But I also wonder how much load many minion would make on Vault then and how often master will do that. I will test and see in practice now - also how good the renewal works.

**Comment by elwood218** [comment]
I didn't have that much time for that topic but anyway had some thoughts and was trying some things.

Which I still don't get is the way how the secretID is used. So if anyone is accessing the Salt-Master then it could just read the secretid and roleid from the config file.. I also don't understand the usage of environment variable. In my eyes it should still work as every Vault Agents works - have a secretid only for startup which is revoked then and then it should renew its token by itself within the process - because anyone could read a file but injecting a process and get the secretid is more complicated. Or do I miss something?

There is also something else but what is not certainly vault related. On every of 1000 or more minion someone could get easy all passwords which are usually not on the filesystem by calling `pillar.items`. Root password, Mysql password and so on. Yes you must be root but being root not necessarily means having access to everything else. But maybe I also miss something here.

**Reply by lkubb** [reply]
I'll try to answer your questions to the best of my ability, but they are quite general.

TLDR: The point is that Salt and the Vault modules are extremely flexible, but you might have to come up with a strategy that satisfies your threat model by yourself.

> Which I still don't get is the way how the secretID is used. So if anyone is accessing the Salt-Master then it could just read the secretid and roleid from the config file..

This is true for any credentials your master config might contain. It should not be world-readable. Anyone that has that level of access to the master would have full access to all minions and could generate arbitrary minion authentication credentials (even if the config was not written to disk) since they would be able to run arbitrary master commands.

Another architectural issue with having a single-use SecretID and caching the token in memory is that some Salt invocations do not have a shared in-memory cache (e.g. a `salt-run` invocation is started in a separate process from the running master daemon). To circumvent that, the token needs to be written to a shared cache anyways, which is usually the disk one in `/var/cache/salt/master`.

> I also don't understand the usage of environment variable. In my eyes it should still work as every Vault Agents works - have a secretid only for startup which is revoked then and then it should renew its token by itself within the process - because anyone could read a file but injecting a process and get the secretid is more complicated. Or do I miss something?

You can use the Vault Agent to manage a token environment variable and query its value via the `env` SDB module, which would then make the master work exactly like the Vault Agent does since it just relies on the Vault Agent output.

Your configuration is possible natively if you account for/can live with the consequences. Specify a single-use secret ID and don't use a `disk` cache. The master daemon then authenticates with the secret ID, making it invalid, and renews the in-memory token as long as it's renewable. It fails once the token becomes invalid or it is restarted without having a fresh SecretID at its disposal.
 
> There is also something else but what is not certainly vault related. On every of 1000 or more minion someone could get easy all passwords which are usually not on the filesystem by calling `pillar.items`. Root password, Mysql password and so on. Yes you must be root but being root not necessarily means having access to everything else. But maybe I also miss something here.

Mostly out of scope in regards to the Vault modules, but this is always an issue when authentication credentials have to be written to disk, i.e. your application does not support Vault natively. Fwiw, I don't include Vault secrets in the pillar, I read them on the minions (in formula map files) directly via `vault.read_secret` / `vault_db.get_creds`. This frankly doesn't make a practical difference, but it means a bit more work for a possible attacker than just running `pillar.items`.

Automation and security levels are a spectrum, which you have to balance according to your threat model.

**Comment by elwood218** [comment]
Hi @lkubb , thank you again for that fast response and help!
So if it has to write the token anywhere and there is no other way then I have to find another way. But unfortunately so far after a time I always get that the secretid is invalid and then the gitfs does not work anymore as I tried to use sdb for getting git password from vault. So it seems like the master does not renew its token. Maybe I forgot a 

I don't know that Vault agent also can provide a environment variable. I only know that it can sink token to file. There is a possibilty with environment variable but I only know that secrets can be set and that the salt-master then needs to be started by the vault agent in supervisor mode.
I also don't know if Salt master would reguarly read the new token created by the agent or if it needs a cronjob which reloads salt master often.

Kind regards
Mathias

**Reply by lkubb** [reply]
> So if it has to write the token anywhere and there is no other way then I have to find another way.

It doesn't **have** to write the token (and it does not when you have not configured a persistent cache), but it can cause some breakage if it does not and the configured single-use SecretID becomes invalid.

> But unfortunately so far after a time I always get that the secretid is invalid and then the gitfs does not work anymore as I tried to use sdb for getting git password from vault. So it seems like the master does not renew its token.

Maybe your Vault traffic is too low. The token is renewed whenever a Vault query is issued (again an architectural issue, Salt extension modules cannot start a background process by themselves). You can [employ an engine](https://github.com/lkubb/salt-vault-formula/blob/master/_engines/vault.py) [this engine would need some import adjustments to work with the saltext] that regularly issues a query to workaround that or just schedule a `vault.query get auth/token/lookup-self` or similar.

Edit: I just remembered there's the [Vault beacon](https://salt-extensions.github.io/saltext-vault/ref/beacons/saltext.vault.beacons.vault_lease.html), which would similarly cause a token renewal when running in the background.

> I don't know that Vault agent also can provide a environment variable. I only know that it can sink token to file. There is a possibilty with environment variable but I only know that secrets can be set and that the salt-master then needs to be started by the vault agent in supervisor mode. I also don't know if Salt master would reguarly read the new token created by the agent or if it needs a cronjob which reloads salt master often.

Tbh I don't use Vault Agent, so I'm out of my comfort zone here. An `sdb://sdbenv/VAULT_TOKEN` should be updated dynamically though, without restarting necessary.

There's a [discussion on Reddit](https://www.reddit.com/r/saltstack/comments/1h04cot/disclosure_of_sensitive_data_via_saltcall/) from today that is related to your questions. If using orchestrations as mentioned there would work for you, you can allow access to your secrets only to the master minion (its minion ID is usually the one configured for the minion running on the same system with a `_master` suffix).
